modules/core/nuxeo-platform-commandline-executor/src/main/java/org/nuxeo/ecm/platform/commandline/executor/service/CommandLineExecutorComponent.java

```diff
@@ -166,58 +164,47 @@ public ExecResult execCommand(String commandName, CmdParameters params) throws C
         if (!availability.isAvailable()) {
             throw new CommandNotAvailable(availability);
         }
-
-        CommandLineDescriptor cmdDesc = commandDescriptors.get(commandName);
-        Executor executor = executors.get(cmdDesc.getExecutor());
-        EnvironmentDescriptor environment = new EnvironmentDescriptor().merge(env).merge(
-                envDescriptors.getOrDefault(commandName, envDescriptors.get(cmdDesc.getCommand())));
-        return executor.exec(cmdDesc, params, environment);
+        var cmdDesc = this.<CommandLineDescriptor> getRegistryContribution(EP_CMD, commandName)
+                          // should not happen
+                          .orElseThrow(() -> new RuntimeException("Command " + commandName + " is not available"));
+        EnvironmentDescriptor environment = new EnvironmentDescriptor().merge(
+                env).merge(envDescriptors.getOrDefault(commandName, envDescriptors.get(cmdDesc.getCommand())));
+        return DEFAULT_EXECUTOR_INSTANCE.exec(cmdDesc, params, environment);
     }
 
     @Override
     public CommandAvailability getCommandAvailability(String commandName) {
-        if (!commandDescriptors.containsKey(commandName)) {
-            return new CommandAvailability(commandName + " is not a registered command");
-        }
-
-        CommandLineDescriptor desc = commandDescriptors.get(commandName);
-        if (desc.isAvailable()) {
-            return new CommandAvailability();
-        } else {
-            return new CommandAvailability(desc.getInstallationDirective(), desc.getInstallErrorMessage());
+        CommandAvailability avail = unavailableCommands.get(commandName);
+        if (avail != null) {
+            return avail;
         }
+        return this.<CommandLineDescriptor> getRegistryContribution(EP_CMD, commandName)
+                   .map(desc -> new CommandAvailability())
+                   .orElse(new CommandAvailability(commandName + " is not a registered command"));
```

Comment:
I agree this old code is quite complicated, and migrating it gave me quite a headache.
The only reason the service stores unavailable commands it to preserve the install message held by descriptors.
I thought it was valuable to keep that information (which is tested), although it may not be used by callers, but since it is wrapped by the CommandNotAvailable exception` #getMessage` method, it is hard to check whether some caller code is relying on this information to be provided...

Reply:
For later